File: web_client/data/mappers/SearchMapper.ts
import {
  SearchFacetOptionDTO,
  SearchFacetsDTO,
  SearchListingDTO,
  SearchPriceRangeFacetDTO,
  SearchResultDTO,
} from '../models/SearchDTO';
import { SearchListing } from '../../domain/models/SearchListing';
import {
  SearchFacetOption,
  SearchFacets,
  SearchPriceRangeFacet,
  SearchResult,
} from '../../domain/models/SearchResult';

export class SearchMapper {
  static listingToDomain(dto: SearchListingDTO): SearchListing {
    return {
      id: dto.id,
      title: dto.title,
      price: dto.price ?? null,
      currency: dto.currency ?? null,
      mediaUrls: dto.media_urls ?? null,
      locationNameRu: dto.location_name_ru ?? null,
      locationNameUz: dto.location_name_uz ?? null,
      refreshedAt: dto.refreshed_at ?? null,
      isPromoted: dto.is_promoted ?? null,
    };
  }

  static resultToDomain(dto: SearchResultDTO): SearchResult {
    const results = Array.isArray(dto.results)
      ? dto.results.map((listing) => this.listingToDomain(listing))
      : dto.results ?? null;

    const facets = this.mapFacets(dto.facets);

    return {
      results,
      total: dto.total ?? null,
      page: dto.page ?? null,
      perPage: dto.per_page ?? null,
      facets,
    };
  }

  private static mapFacets(facets?: SearchFacetsDTO | null): SearchFacets | null | undefined {
    if (facets === undefined) return undefined;
    if (facets === null) return null;

    const mapped: SearchFacets = {};

    const categories = this.mapFacetOptions(facets.categories);
    if (categories !== undefined) {
      mapped.categories = categories;
    }

    const locations = this.mapFacetOptions(facets.locations);
    if (locations !== undefined) {
      mapped.locations = locations;
    }

    const conditions = this.mapFacetOptions(facets.conditions);
    if (conditions !== undefined) {
      mapped.conditions = conditions;
    }

    const priceRange = this.mapPriceRange(facets.price_range);
    if (priceRange !== undefined) {
      mapped.price_range = priceRange;
    }

    const attributes = this.mapFacetAttributes(facets.attributes);
    if (attributes !== undefined) {
      mapped.attributes = attributes;
    }

    return mapped;
  }

  private static mapFacetOptions(
    options?: SearchFacetOptionDTO[] | null,
  ): SearchFacetOption[] | null | undefined {
    if (options === undefined) return undefined;
    if (options === null) return null;

    return options.map((option) => ({
      key: option?.key ?? null,
      count: option?.count ?? null,
    }));
  }

  private static mapPriceRange(
    priceRange?: SearchPriceRangeFacetDTO | null,
  ): SearchPriceRangeFacet | null | undefined {
    if (priceRange === undefined) return undefined;
    if (priceRange === null) return null;

    return {
      min: priceRange.min ?? null,
      max: priceRange.max ?? null,
      currency: priceRange.currency ?? null,
    };
  }

  private static mapFacetAttributes(
    attributes?: Record<string, SearchFacetOptionDTO[] | null> | null,
  ): Record<string, SearchFacetOption[] | null> | null | undefined {
    if (attributes === undefined) return undefined;
    if (attributes === null) return null;

    const mapped: Record<string, SearchFacetOption[] | null> = {};

    Object.entries(attributes).forEach(([key, value]) => {
      if (value === undefined) {
        return;
      }
      const options = value === null ? null : this.mapFacetOptions(value) ?? null;
      mapped[key] = options;
    });

    return mapped;
  }
}

File: mobile_client/lib/data/mappers/search_mapper.dart
/// Search mappers for SearchResultDto ↔ SearchResult conversion.
library;

import '../../domain/entities/search_result.dart';
import '../../domain/entities/user.dart';
import '../models/search/search_dto.dart';
import '../models/listing/listing_dto.dart';

/// Mapper for ListingSellerDto → ListingSeller (for search results).
extension SearchListingSellerDtoMapper on ListingSellerDto {
  ListingSeller toEntity() {
    return ListingSeller(
      id: id,
      name: name,
      avatarUrl: avatarUrl,
      since: since,
      logo: logo,
      banner: banner,
      phone: phone,
      lastActiveAt: lastActiveAt,
    );
  }
}

/// Mapper for SearchListingDto → SearchListing.
extension SearchListingDtoMapper on SearchListingDto {
  SearchListing toEntity() {
    return SearchListing(
      id: id,
      title: title,
      price: price,
      description: description,
      condition: condition,
      qualityScore: qualityScore,
      score: score,
      currency: currency,
      mediaUrls: mediaUrls,
      locationNameRu: locationNameRu,
      locationNameUz: locationNameUz,
      refreshedAt: refreshedAt,
      isPromoted: isPromoted,
      seller: seller?.toEntity(),
    );
  }
}

/// Mapper for SearchFacetOptionDto → SearchFacetOption.
extension SearchFacetOptionDtoMapper on SearchFacetOptionDto {
  SearchFacetOption toEntity() {
    return SearchFacetOption(
      key: key,
      count: count,
    );
  }
}

/// Mapper for SearchPriceRangeFacetDto → SearchPriceRangeFacet.
extension SearchPriceRangeFacetDtoMapper on SearchPriceRangeFacetDto {
  SearchPriceRangeFacet toEntity() {
    return SearchPriceRangeFacet(
      min: min,
      max: max,
      currency: currency,
    );
  }
}

/// Mapper for SearchFacetsDto → SearchFacets.
extension SearchFacetsDtoMapper on SearchFacetsDto {
  SearchFacets toEntity() {
    return SearchFacets(
      categories: categories?.map((f) => f.toEntity()).toList(),
      locations: locations?.map((f) => f.toEntity()).toList(),
      conditions: conditions?.map((f) => f.toEntity()).toList(),
      priceRange: priceRange?.toEntity(),
      attributes: attributes?.map(
        (key, value) => MapEntry(
          key,
          value?.map((f) => f.toEntity()).toList() ?? [],
        ),
      ),
    );
  }
}

/// Mapper for SearchResultDto → SearchResult.
extension SearchResultDtoMapper on SearchResultDto {
  SearchResult toEntity() {
    return SearchResult(
      results: results?.map((r) => r.toEntity()).toList() ?? [],
      total: total ?? 0,
      page: page ?? 1,
      perPage: perPage ?? 24,
      facets: facets?.toEntity(),
    );
  }
}

/// Mapper for SearchParams → query parameters map.
extension SearchParamsMapper on SearchParams {
  Map<String, dynamic> toQueryParams() {
    final params = <String, dynamic>{};
    
    if (query != null && query!.isNotEmpty) params['q'] = query;
    if (category != null) params['category'] = category.toString();
    if (location != null) params['location'] = location.toString();
    if (priceMin != null) params['price_min'] = priceMin.toString();
    if (priceMax != null) params['price_max'] = priceMax.toString();
    if (condition != null) params['condition'] = condition;
    if (sort != null) params['sort'] = sort;
    params['page'] = page.toString();
    params['per_page'] = perPage.toString();
    
    // Add dynamic attributes
    attributes?.forEach((key, value) {
      if (value != null) {
        params['attr_$key'] = value.toString();
      }
    });
    
    return params;
  }
}
